File: 0c-bason/assignment-1/cpp/src/json_converter.cpp
#include "json_converter.h"
#include "ron64.h"

#include <sstream>
#include <stdexcept>
#include <cctype>

////////////////////////////////////////////////////////////////////////////////

namespace NBason {

////////////////////////////////////////////////////////////////////////////////

namespace {

class TJsonParser {
public:
    TJsonParser(const std::string& json)
        : Json_(json)
        , Pos_(0)
    { }

    TBasonRecord Parse()
    {
        SkipWhitespace();
        return ParseValue("");
    }

private:
    const std::string& Json_;
    size_t Pos_;

    void SkipWhitespace()
    {
        while (Pos_ < Json_.size() && std::isspace(Json_[Pos_])) {
            ++Pos_;
        }
    }

    char Peek()
    {
        SkipWhitespace();
        if (Pos_ >= Json_.size()) {
            throw std::runtime_error("Unexpected end of JSON");
        }
        return Json_[Pos_];
    }

    char Next()
    {
        SkipWhitespace();
        if (Pos_ >= Json_.size()) {
            throw std::runtime_error("Unexpected end of JSON");
        }
        return Json_[Pos_++];
    }

    void Expect(char ch)
    {
        if (Next() != ch) {
            throw std::runtime_error(std::string("Expected '") + ch + "'");
        }
    }

    std::string ParseString()
    {
        Expect('"');
        std::string result;
        
        while (Pos_ < Json_.size() && Json_[Pos_] != '"') {
            if (Json_[Pos_] == '\\') {
                ++Pos_;
                if (Pos_ >= Json_.size()) {
                    throw std::runtime_error("Unterminated string escape");
                }
                char escaped = Json_[Pos_++];
                switch (escaped) {
                    case '"': result += '"'; break;
                    case '\\': result += '\\'; break;
                    case '/': result += '/'; break;
                    case 'b': result += '\b'; break;
                    case 'f': result += '\f'; break;
                    case 'n': result += '\n'; break;
                    case 'r': result += '\r'; break;
                    case 't': result += '\t'; break;
                    case 'u':
                        // Unicode escape - simplified handling
                        throw std::runtime_error("Unicode escapes not fully supported");
                    default:
                        throw std::runtime_error("Invalid escape sequence");
                }
            } else {
                result += Json_[Pos_++];
            }
        }
        
        Expect('"');
        return result;
    }

    std::string ParseNumber()
    {
        size_t start = Pos_;
        
        if (Json_[Pos_] == '-') {
            ++Pos_;
        }
        
        if (Pos_ >= Json_.size() || !std::isdigit(Json_[Pos_])) {
            throw std::runtime_error("Invalid number");
        }
        
        while (Pos_ < Json_.size() && 
               (std::isdigit(Json_[Pos_]) || Json_[Pos_] == '.' || 
                Json_[Pos_] == 'e' || Json_[Pos_] == 'E' || 
                Json_[Pos_] == '+' || Json_[Pos_] == '-')) {
            ++Pos_;
        }
        
        return Json_.substr(start, Pos_ - start);
    }

    TBasonRecord ParseValue(const std::string& key)
    {
        char ch = Peek();
        
        if (ch == '"') {
            return TBasonRecord(EBasonType::String, key, ParseString());
        } else if (ch == '{') {
            return ParseObject(key);
        } else if (ch == '[') {
            return ParseArray(key);
        } else if (ch == 't' || ch == 'f' || ch == 'n') {
            return ParseLiteral(key);
        } else if (ch == '-' || std::isdigit(ch)) {
            return TBasonRecord(EBasonType::Number, key, ParseNumber());
        }
        
        throw std::runtime_error("Unexpected character in JSON");
    }

    TBasonRecord ParseLiteral(const std::string& key)
    {
        if (Json_.substr(Pos_, 4) == "true") {
            Pos_ += 4;
            return TBasonRecord(EBasonType::Boolean, key, "true");
        } else if (Json_.substr(Pos_, 5) == "false") {
            Pos_ += 5;
            return TBasonRecord(EBasonType::Boolean, key, "false");
        } else if (Json_.substr(Pos_, 4) == "null") {
            Pos_ += 4;
            return TBasonRecord(EBasonType::Boolean, key, "");  // null = empty value
        }
        
        throw std::runtime_error("Invalid literal");
    }

    TBasonRecord ParseObject(const std::string& key)
    {
        TBasonRecord record(EBasonType::Object, key, "");
        Expect('{');
        
        SkipWhitespace();
        if (Peek() == '}') {
            Next();
            return record;
        }
        
        while (true) {
            std::string fieldName = ParseString();
            Expect(':');
            auto value = ParseValue(fieldName);
            record.Children.push_back(std::move(value));
            
            SkipWhitespace();
            if (Peek() == '}') {
                Next();
                break;
            }
            Expect(',');
        }
        
        return record;
    }

    TBasonRecord ParseArray(const std::string& key)
    {
        TBasonRecord record(EBasonType::Array, key, "");
        Expect('[');
        
        SkipWhitespace();
        if (Peek() == ']') {
            Next();
            return record;
        }
        
        size_t index = 0;
        while (true) {
            std::string indexKey = EncodeRon64(index++);
            auto value = ParseValue(indexKey);
            record.Children.push_back(std::move(value));
            
            SkipWhitespace();
            if (Peek() == ']') {
                Next();
                break;
            }
            Expect(',');
        }
        
        return record;
    }
};

void RecordToJson(const TBasonRecord& record, std::ostringstream& out)
{
    switch (record.Type) {
        case EBasonType::String:
            out << '"';
            for (char ch : record.Value) {
                switch (ch) {
                    case '"': out << "\\\""; break;
                    case '\\': out << "\\\\"; break;
                    case '\b': out << "\\b"; break;
                    case '\f': out << "\\f"; break;
                    case '\n': out << "\\n"; break;
                    case '\r': out << "\\r"; break;
                    case '\t': out << "\\t"; break;
                    default:
                        if (ch < 32) {
                            // Control character - escape as \uXXXX
                            out << "\\u00" << std::hex << (int)(unsigned char)ch << std::dec;
                        } else {
                            out << ch;
                        }
                }
            }
            out << '"';
            break;

        case EBasonType::Number:
            out << record.Value;
            break;

        case EBasonType::Boolean:
            if (record.Value.empty()) {
                out << "null";
            } else {
                out << record.Value;
            }
            break;

        case EBasonType::Object:
            out << '{';
            for (size_t i = 0; i < record.Children.size(); ++i) {
                if (i > 0) out << ',';
                out << '"' << record.Children[i].Key << "\":";
                RecordToJson(record.Children[i], out);
            }
            out << '}';
            break;

        case EBasonType::Array:
            out << '[';
            for (size_t i = 0; i < record.Children.size(); ++i) {
                if (i > 0) out << ',';
                RecordToJson(record.Children[i], out);
            }
            out << ']';
            break;
    }
}

} // namespace

////////////////////////////////////////////////////////////////////////////////

std::vector<uint8_t> JsonToBason(const std::string& json)
{
    TJsonParser parser(json);
    TBasonRecord record = parser.Parse();
    return EncodeBason(record);
}

std::string BasonToJson(const uint8_t* data, size_t len)
{
    auto [record, _] = DecodeBason(data, len);
    
    std::ostringstream out;
    RecordToJson(record, out);
    return out.str();
}

////////////////////////////////////////////////////////////////////////////////

} // namespace NBason

File: 0c-bason/assignment-1/go/pkg/bason/json_converter.go
package bason

import (
	"bytes"
	"errors"
	"unicode"
)

var (
	ErrUnexpectedEnd = errors.New("unexpected end of JSON")
	ErrExpectedChar  = errors.New("expected character")
	ErrInvalidEscape = errors.New("invalid escape sequence")
	ErrUnicodeEscape = errors.New("unicode escapes not fully supported")
	ErrInvalidNumber = errors.New("invalid number")
	ErrInvalidLiteral = errors.New("invalid literal")
	ErrUnexpectedChar = errors.New("unexpected character in JSON")
)

type jsonParser struct {
	json []byte
	pos  int
}

func newJSONParser(json string) *jsonParser {
	return &jsonParser{json: []byte(json)}
}

func (p *jsonParser) skipWhitespace() {
	for p.pos < len(p.json) && unicode.IsSpace(rune(p.json[p.pos])) {
		p.pos++
	}
}

func (p *jsonParser) peek() (byte, error) {
	p.skipWhitespace()
	if p.pos >= len(p.json) {
		return 0, ErrUnexpectedEnd
	}
	return p.json[p.pos], nil
}

func (p *jsonParser) next() (byte, error) {
	p.skipWhitespace()
	if p.pos >= len(p.json) {
		return 0, ErrUnexpectedEnd
	}
	b := p.json[p.pos]
	p.pos++
	return b, nil
}

func (p *jsonParser) expect(ch byte) error {
	b, err := p.next()
	if err != nil {
		return err
	}
	if b != ch {
		return ErrExpectedChar
	}
	return nil
}

func (p *jsonParser) parseString() (string, error) {
	if err := p.expect('"'); err != nil {
		return "", err
	}
	var result []byte
	for p.pos < len(p.json) && p.json[p.pos] != '"' {
		if p.json[p.pos] == '\\' {
			p.pos++
			if p.pos >= len(p.json) {
				return "", ErrInvalidEscape
			}
			switch p.json[p.pos] {
			case '"':
				result = append(result, '"')
			case '\\':
				result = append(result, '\\')
			case '/':
				result = append(result, '/')
			case 'b':
				result = append(result, '\b')
			case 'f':
				result = append(result, '\f')
			case 'n':
				result = append(result, '\n')
			case 'r':
				result = append(result, '\r')
			case 't':
				result = append(result, '\t')
			case 'u':
				return "", ErrUnicodeEscape
			default:
				return "", ErrInvalidEscape
			}
			p.pos++
		} else {
			result = append(result, p.json[p.pos])
			p.pos++
		}
	}
	if err := p.expect('"'); err != nil {
		return "", err
	}
	return string(result), nil
}

func (p *jsonParser) parseNumber() (string, error) {
	start := p.pos
	if p.json[p.pos] == '-' {
		p.pos++
	}
	if p.pos >= len(p.json) || !isDigit(p.json[p.pos]) {
		return "", ErrInvalidNumber
	}
	for p.pos < len(p.json) {
		b := p.json[p.pos]
		if isDigit(b) || b == '.' || b == 'e' || b == 'E' || b == '+' || b == '-' {
			p.pos++
		} else {
			break
		}
	}
	return string(p.json[start:p.pos]), nil
}

func (p *jsonParser) parseValue(key string) (Record, error) {
	ch, err := p.peek()
	if err != nil {
		return Record{}, err
	}
	switch ch {
	case '"':
		s, err := p.parseString()
		if err != nil {
			return Record{}, err
		}
		return Record{Type: TypeString, Key: key, Value: s}, nil
	case '{':
		return p.parseObject(key)
	case '[':
		return p.parseArray(key)
	case 't', 'f', 'n':
		return p.parseLiteral(key)
	case '-':
		fallthrough
	default:
		if isDigit(ch) {
			n, err := p.parseNumber()
			if err != nil {
				return Record{}, err
			}
			return Record{Type: TypeNumber, Key: key, Value: n}, nil
		}
		return Record{}, ErrUnexpectedChar
	}
}

func (p *jsonParser) parseLiteral(key string) (Record, error) {
	if p.pos+4 <= len(p.json) && bytes.Equal(p.json[p.pos:p.pos+4], []byte("true")) {
		p.pos += 4
		return Record{Type: TypeBoolean, Key: key, Value: "true"}, nil
	}
	if p.pos+5 <= len(p.json) && bytes.Equal(p.json[p.pos:p.pos+5], []byte("false")) {
		p.pos += 5
		return Record{Type: TypeBoolean, Key: key, Value: "false"}, nil
	}
	if p.pos+4 <= len(p.json) && bytes.Equal(p.json[p.pos:p.pos+4], []byte("null")) {
		p.pos += 4
		return Record{Type: TypeBoolean, Key: key, Value: ""}, nil
	}
	return Record{}, ErrInvalidLiteral
}

func (p *jsonParser) parseObject(key string) (Record, error) {
	record := Record{Type: TypeObject, Key: key}
	if err := p.expect('{'); err != nil {
		return Record{}, err
	}
	p.skipWhitespace()
	ch, _ := p.peek()
	if ch == '}' {
		p.next()
		return record, nil
	}
	for {
		fieldName, err := p.parseString()
		if err != nil {
			return Record{}, err
		}
		if err := p.expect(':'); err != nil {
			return Record{}, err
		}
		child, err := p.parseValue(fieldName)
		if err != nil {
			return Record{}, err
		}
		record.Children = append(record.Children, child)
		p.skipWhitespace()
		ch, _ = p.peek()
		if ch == '}' {
			p.next()
			return record, nil
		}
		if err := p.expect(','); err != nil {
			return Record{}, err
		}
	}
}

func (p *jsonParser) parseArray(key string) (Record, error) {
	record := Record{Type: TypeArray, Key: key}
	if err := p.expect('['); err != nil {
		return Record{}, err
	}
	p.skipWhitespace()
	ch, _ := p.peek()
	if ch == ']' {
		p.next()
		return record, nil
	}
	index := uint64(0)
	for {
		indexKey := EncodeRon64(index)
		index++
		child, err := p.parseValue(indexKey)
		if err != nil {
			return Record{}, err
		}
		record.Children = append(record.Children, child)
		p.skipWhitespace()
		ch, _ = p.peek()
		if ch == ']' {
			p.next()
			return record, nil
		}
		if err := p.expect(','); err != nil {
			return Record{}, err
		}
	}
}

func (p *jsonParser) parse() (Record, error) {
	p.skipWhitespace()
	return p.parseValue("")
}

func recordToJSON(record Record, out *bytes.Buffer) {
	switch record.Type {
	case TypeString:
		out.WriteByte('"')
		for i := 0; i < len(record.Value); i++ {
			ch := record.Value[i]
			switch ch {
			case '"':
				out.WriteString(`\"`)
			case '\\':
				out.WriteString(`\\`)
			case '\b':
				out.WriteString(`\b`)
			case '\f':
				out.WriteString(`\f`)
			case '\n':
				out.WriteString(`\n`)
			case '\r':
				out.WriteString(`\r`)
			case '\t':
				out.WriteString(`\t`)
			default:
				if ch < 32 {
					out.WriteString(`\u00`)
					out.WriteByte("0123456789abcdef"[ch>>4])
					out.WriteByte("0123456789abcdef"[ch&0x0f])
				} else {
					out.WriteByte(ch)
				}
			}
		}
		out.WriteByte('"')
	case TypeNumber:
		out.WriteString(record.Value)
	case TypeBoolean:
		if record.Value == "" {
			out.WriteString("null")
		} else {
			out.WriteString(record.Value)
		}
	case TypeObject:
		out.WriteByte('{')
		for i, child := range record.Children {
			if i > 0 {
				out.WriteByte(',')
			}
			out.WriteByte('"')
			out.WriteString(child.Key)
			out.WriteString(`":`)
			recordToJSON(child, out)
		}
		out.WriteByte('}')
	case TypeArray:
		out.WriteByte('[')
		for i, child := range record.Children {
			if i > 0 {
				out.WriteByte(',')
			}
			recordToJSON(child, out)
		}
		out.WriteByte(']')
	}
}

// JsonToBason converts JSON text to a BASON byte stream (nested mode).
func JsonToBason(json string) ([]byte, error) {
	parser := newJSONParser(json)
	record, err := parser.parse()
	if err != nil {
		return nil, err
	}
	return EncodeBason(record)
}

// BasonToJson converts a BASON byte stream to JSON text.
func BasonToJson(data []byte) (string, error) {
	record, _, err := DecodeBason(data)
	if err != nil {
		return "", err
	}
	var out bytes.Buffer
	recordToJSON(record, &out)
	return out.String(), nil
}
